src/math/power.c

```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 Intel Corporation. All rights reserved.
+//
+// Author: Shriram Shastry <malladi.sastry@linux.intel.com>
+//
+//
+
+ /* Include Files */
+#include <sof/math/power.h>
+
+/* Function Declarations */
+static int32_t div_s32s64(int64_t denominator);
+
+/* Function Definitions */
+/**
+ * Arguments	: int64_t denominator
+ *
+ * Return Type	: int32_t
+ */
+static inline int32_t div_s32s64(int64_t denominator)
+{
+	uint64_t b_denominator;
+	int32_t quotient;
+
+	if (denominator == 0) {
+		quotient = INT32_MAX;
+	} else {
+		if (denominator < 0)
+			b_denominator = ~(uint64_t)denominator + 1;
+		else
+			b_denominator = (uint64_t)denominator;
+
+		b_denominator = 1125899906842624ULL / b_denominator;
+		if (denominator < 0L)
+			quotient = (int32_t)-(int64_t)b_denominator;
+		else
+			quotient = (int32_t)b_denominator;
+	}
+	return quotient;
+}
+
+/**
+ * Arguments	: int32_t b
+ * base = [-1 to 2^5];Q7.25
+ *
+ *		: int32_t e
+ * exponent = [-1 to 3];Q3.29
+ *
+ * Return Type	: int32_t
+ * power = [-32768 to +32768];Q17.15
+ */
+int32_t power_scalar_function(int32_t b, int32_t e)
+{
+	int32_t i;
+	int32_t k;
+	int32_t multiplier;
+	int32_t p;
+
+	p = 32768;
+
+	if (e < 0) {
+		if (e <= INT32_MIN)
+			e = INT32_MAX;
+		else
+			e = -e;
+
+		if (b == 0)
+			multiplier = INT32_MAX;
+		else
+			multiplier = div_s32s64((int64_t)b);
+
+	} else {
+		multiplier = b;
+	}
+	i = e >> 29;
+	for (k = 0; k < i; k++)
+		p = (int32_t)(((int64_t)p * (int64_t)multiplier) >> 25);
```

Comment:
this would probably be much simpler if you left p as i64 then just cast it at the return

Reply:
somehow I didn't fully understand your suggestion. Can you please help me with more ?

Reply:
this just spreads out the complexity rather than having a pile of casts all on one line
```
int64_t multiplier;
int64_t p = MAX_LIMIT;

...

for (k = 0; k < i; k++)
		p = (int32_t)((p * multiplier) >> 25);
return (int32_t)p;
```

Reply:
I think you over looked. `(int64_t)multiplier` is needed, otherwise return value containing MSB signed bit is lost. Therefore we loose accuracy.

Test -2 and Test-4 fails

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/shastry/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/shastry/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<!--[if !mso]>
<style>
v\:* {behavior:url(#default#VML);}
o\:* {behavior:url(#default#VML);}
x\:* {behavior:url(#default#VML);}
.shape {behavior:url(#default#VML);}
</style>
<![endif]-->
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding:0px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{text-align:center;
	vertical-align:middle;
	border:.5pt solid windowtext;
	white-space:normal;}
.xl66
	{color:#9C0006;
	text-align:center;
	vertical-align:middle;
	border:.5pt solid windowtext;
	background:#FFC7CE;
	mso-pattern:black none;
	white-space:normal;}
.xl67
	{text-align:center;
	vertical-align:middle;
	border-top:.5pt solid windowtext;
	border-right:.5pt solid windowtext;
	border-bottom:.5pt solid windowtext;
	border-left:1.0pt solid windowtext;
	white-space:normal;}
.xl68
	{text-align:center;
	vertical-align:middle;
	border-top:.5pt solid windowtext;
	border-right:1.0pt solid windowtext;
	border-bottom:.5pt solid windowtext;
	border-left:.5pt solid windowtext;
	white-space:normal;}
.xl69
	{text-align:center;
	vertical-align:middle;
	border-top:.5pt solid windowtext;
	border-right:.5pt solid windowtext;
	border-bottom:1.0pt solid windowtext;
	border-left:1.0pt solid windowtext;
	white-space:normal;}
.xl70
	{text-align:center;
	vertical-align:middle;
	border-top:.5pt solid windowtext;
	border-right:.5pt solid windowtext;
	border-bottom:1.0pt solid windowtext;
	border-left:.5pt solid windowtext;
	white-space:normal;}
.xl71
	{color:#9C0006;
	text-align:center;
	vertical-align:middle;
	border-top:.5pt solid windowtext;
	border-right:1.0pt solid windowtext;
	border-bottom:1.0pt solid windowtext;
	border-left:.5pt solid windowtext;
	background:#FFC7CE;
	mso-pattern:black none;
	white-space:normal;}
.xl72
	{color:#006100;
	text-align:center;
	vertical-align:middle;
	border-top:.5pt solid windowtext;
	border-right:1.0pt solid windowtext;
	border-bottom:.5pt solid windowtext;
	border-left:.5pt solid windowtext;
	background:#C6EFCE;
	mso-pattern:black none;
	white-space:normal;}
.xl73
	{font-weight:700;
	text-align:center;
	vertical-align:middle;
	border-top:.5pt solid windowtext;
	border-right:.5pt solid windowtext;
	border-bottom:.5pt solid windowtext;
	border-left:1.0pt solid windowtext;
	white-space:normal;}
.xl74
	{font-weight:700;
	text-align:center;
	vertical-align:middle;
	border-top:none;
	border-right:.5pt solid windowtext;
	border-bottom:.5pt solid windowtext;
	border-left:1.0pt solid windowtext;
	white-space:normal;}
.xl75
	{color:#006100;
	text-align:center;
	vertical-align:middle;
	border-top:none;
	border-right:.5pt solid windowtext;
	border-bottom:.5pt solid windowtext;
	border-left:.5pt solid windowtext;
	background:#C6EFCE;
	mso-pattern:black none;
	white-space:normal;}
.xl76
	{text-align:center;
	vertical-align:middle;
	border-top:none;
	border-right:.5pt solid windowtext;
	border-bottom:.5pt solid windowtext;
	border-left:.5pt solid windowtext;
	white-space:normal;}
.xl77
	{text-align:center;
	vertical-align:middle;
	border-top:none;
	border-right:1.0pt solid windowtext;
	border-bottom:.5pt solid windowtext;
	border-left:.5pt solid windowtext;
	white-space:normal;}
.xl78
	{font-weight:700;
	text-align:center;
	vertical-align:middle;
	border-top:1.0pt solid windowtext;
	border-right:.5pt solid windowtext;
	border-bottom:1.0pt solid windowtext;
	border-left:1.0pt solid windowtext;
	white-space:normal;}
.xl79
	{font-weight:700;
	text-align:center;
	vertical-align:middle;
	border-top:1.0pt solid windowtext;
	border-right:.5pt solid windowtext;
	border-bottom:1.0pt solid windowtext;
	border-left:.5pt solid windowtext;
	white-space:normal;}
.xl80
	{font-weight:700;
	text-align:center;
	vertical-align:middle;
	border-top:1.0pt solid windowtext;
	border-right:1.0pt solid windowtext;
	border-bottom:1.0pt solid windowtext;
	border-left:.5pt solid windowtext;
	white-space:normal;}
-->
</style>
</head>

<body link="#0563C1" vlink="#954F72">



case   negative denominator | code | variables | data type | initial data   value | update p value
-- | -- | -- | -- | -- | --
TEST-1 | p =   (int32_t)(((int64_t)p * (int64_t)multiplier) >> 25); |   |   |   |  
  |   | e | int | 0x20000000 |  
  |   | i | int | 0x00000001 |  
  |   | k | int | 0x00000000 |  
  |   | multiplier | int | 0xfe000000 |  
  |   | p | int | 0x00008000 | 0xffff8000
TEST-2 | p   = (int32_t)((p * multiplier) >> 25); |   |   |   |  
  |   | e | int | 0x20000000 |  
  |   | i | int | 0x00000001 |  
  |   | k | int | 0x00000000 |  
  |   | multiplier | int | 0xfe000000 |  
  |   | p | int | 0x00008000 | 0x00000000
case positive denominator | code | variables | data   type | initial   data value | update   p value
TEST-3 | p =   (int32_t)(((int64_t)p * (int64_t)multiplier) >> 25); |   |   |   |  
  |   | e | int | 0x20000000 |  
  |   | i | int | 0x00000001 |  
  |   | k | int | 0x00000000 |  
  |   | multiplier | int | 0x02000000 |  
  |   | p | int | 0x00008000 | 0x00008000
TEST-4 | p   = (int32_t)((p * multiplier) >> 25); |   |   |   |  
  |   | e | int | 0x20000000 |  
  |   | i | int | 0x00000001 |  
  |   | k | int | 0x00000000 |  
  |   | multiplier | int | 0x02000000 |  
  |   | p | int | 0x00008000 | 0x00000000



</body>

</html>

We are fine `p = (int32_t)((p * multiplier) >> 25);`